python/tvm/build_module.py

```diff
@@ -68,6 +72,68 @@ def __exit__(self, ptype, value, trace):
 
 BuildConfig.current = BuildConfig()
 
+class DumpIR(object):
+    '''dump ir for each pass'''
+    pass_id = 0
+    decorated = False
+    def __init__(self):
+        self._local_decorated = False
+
+    def decorate(self, func):
+        ''' decorate the pass function'''
+        def dump(*args, **kwargs):
+            '''dump function'''
+            retv = func(*args, **kwargs)
+            if not isinstance(retv, (_stmt.Stmt, container.LoweredFunc, container.Array)):
+                return retv
+            pname = str(DumpIR.pass_id) + "_" + func.func_name + "_ir.cc"
+            with open(pname, "a") as f:
+                out = retv.body if isinstance(retv, container.LoweredFunc) else retv
+                f.write(str(out))
+                if isinstance(retv, container.Array):
+                    for x in retv:
+                        out = x.body if isinstance(x, container.LoweredFunc) else x
+                        f.write("---------", x.name, "\n", str(out), "\n-----------\n")
+                DumpIR.pass_id += 1
+            return retv
+        return dump
+
+    def decorate_irpass(self):
+        for k, v in vars(ir_pass).items():
+            vars(ir_pass)[k] = self.decorate(v) if isinstance(v, types.FunctionType) else v
+        schedule.ScheduleOps = self.decorate(schedule.ScheduleOps)
+
+    def decorate_custompass(self):
+        cfg = BuildConfig.current
+        add_lower_pass = cfg.add_lower_pass if cfg.add_lower_pass else []
+        pass_list = [(x[0], self.decorate(x[1])) for x in add_lower_pass]
+        BuildConfig.current.add_lower_pass = pass_list
+
+    def __enter__(self, pass_inp=None):
+        if DumpIR.decorated is True:
+            return
+        self._old_irpass = vars(ir_pass).items()
+        self._old_sgpass = schedule.ScheduleOps
+        self._old_custom_pass = BuildConfig.current.add_lower_pass
+        self.decorate_irpass()
+        self.decorate_custompass()
+        self._local_decorated = True
+        DumpIR.decorated = True
+        DumpIR.pass_id = 0
+
+    def __exit__(self, ptype, value, trace):
+        if not (self._local_decorated and DumpIR.decorated):
```

Comment:
there is no need additional checks, just allow recursive decoration if needed

Reply:
when lower nested in build, you don't want to decorate twice

Reply:
Add a global counter then for decorated level, and only do decoration at level 1